src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -1253,27 +1257,39 @@ void ShenandoahFreeSet::reserve_regions(size_t to_reserve, size_t to_reserve_old
     size_t idx = i - 1;
     ShenandoahHeapRegion* r = _heap->get_region(idx);
-    if (_free_sets.in_free_set(idx, Mutator)) {
-      assert (!r->is_old(), "mutator_is_free regions should not be affiliated OLD");
-      size_t ac = alloc_capacity(r);
-      assert (ac > 0, "Membership in free set implies has capacity");
-
-      // OLD regions that have available memory are already in the old_collector free set
-      if ((_free_sets.capacity_of(OldCollector) < to_reserve_old) && (r->is_trash() || !r->is_affiliated())) {
-        _free_sets.move_to_set(idx, OldCollector, alloc_capacity(r));
+    if (!_free_sets.in_free_set(idx, Mutator)) {
+      continue;
+    }
+
+    size_t ac = alloc_capacity(r);
+    assert (ac > 0, "Membership in free set implies has capacity");
+    assert (!r->is_old(), "mutator_is_free regions should not be affiliated OLD");
+
+    bool move_to_old = _free_sets.capacity_of(OldCollector) < to_reserve_old;
+    bool move_to_young = _free_sets.capacity_of(Collector) < to_reserve;
+
+    if (!move_to_old && !move_to_young) {
+      // We've satisfied both to_reserve and to_reserved_old
+      break;
+    }
+
+    if (move_to_old) {
+      if (r->is_trash() || !r->is_affiliated()) {
+        // OLD regions that have available memory are already in the old_collector free set
+        _free_sets.move_to_set(idx, OldCollector, ac);
         log_debug(gc, free)("  Shifting region " SIZE_FORMAT " from mutator_free to old_collector_free", idx);
-      } else if (_free_sets.capacity_of(Collector) < to_reserve) {
-        // Note: In a previous implementation, regions were only placed into the survivor space (collector_is_free) if
-        // they were entirely empty.  I'm not sure I understand the rational for that.  That alternative behavior would
-        // tend to mix survivor objects with ephemeral objects, making it more difficult to reclaim the memory for the
-        // ephemeral objects.  It also delays aging of regions, causing promotion in place to be delayed.
-        _free_sets.move_to_set(idx, Collector, ac);
-        log_debug(gc)("  Shifting region " SIZE_FORMAT " from mutator_free to collector_free", idx);
-      } else {
-        // We've satisfied both to_reserve and to_reserved_old
-        break;
+        continue;
       }
     }
+
+    if (move_to_young) {
+      // Note: In a previous implementation, regions were only placed into the survivor space (collector_is_free) if
+      // they were entirely empty.  I'm not sure I understand the rationale for that.  That alternative behavior would
+      // tend to mix survivor objects with ephemeral objects, making it more difficult to reclaim the memory for the
+      // ephemeral objects.  It also delays aging of regions, causing promotion in place to be delayed.
+      _free_sets.move_to_set(idx, Collector, ac);
+      log_debug(gc)("  Shifting region " SIZE_FORMAT " from mutator_free to collector_free", idx);
+    }
   }
 }
 
```

Comment:
When we rebuild, we compute target amounts of memory to_reserve_old and to_reserve_young.  We may not satisfy our targets if the Mutator.is_free memory is "too fragmented".  This would be rare, I think, but maybe deserving of a log_gc(info) message whenever we get to the end of rebuild and we are short.  I think the rest of the implementation is supposed to be robust to this scenario.  (i.e. if we did not reserve as much as we wanted to, then we will target a smaller collection set (in old or young)).